src/identities/mnemonic.ts

```
@@ -211,4 +237,127 @@ export default class Mnemonic extends Identity implements IdentityInterface {
+
+    const incrementOrResetCounter = (addresses: AddressInterfaceExtended[]) => (
+      hasBeenUsed: boolean,
+      index: number
+    ) => {
+      if (hasBeenUsed === true) {
+        counter = 0;
+        restoredAddresses.push(addresses[index]);
+      } else {
+        counter += 1;
+      }
+    };
+
+    while (counter < NOT_USED_ADDRESSES_LIMIT) {
+      // generate addresses to test
+      const addressesToTest = await this.generateSetOfAddresses(
+        index,
+        NOT_USED_ADDRESSES_LIMIT - counter
+      );
+      // test all addresses asynchronously using restorer.
+      const hasBeenUsedArray: boolean[] = await this.checkAddressesWithRestorer(
+        addressesToTest
+      );
+
+      // iterate through array
+      // if address has been used before = push to restoredAddresses array
+      // else increment counter
+      hasBeenUsedArray.forEach(incrementOrResetCounter(addressesToTest));
+      index += NOT_USED_ADDRESSES_LIMIT;
+    }
+
+    // check for change address
+    const changeAddresses: AddressInterfaceExtended[] = await Promise.all(
+      restoredAddresses.map(addr => this.addressToChangeAddressAsync(addr))
+    );
+
+    const hasBeenUsedArrayChange: boolean[] = await this.checkAddressesWithRestorer(
+      changeAddresses
+    );
+
+    hasBeenUsedArrayChange.forEach((hasBeenUsed: boolean, index: number) => {
+      if (hasBeenUsed) {
+        restoredAddresses.push(changeAddresses[index]);
+      }
+    });
+
+    // return the restored address
+    return restoredAddresses;
+  }
+
+  /**
+   * Restore try to (1) generate and verify a range of addresses & (2) persist the address to the instance cache.
+   * Then it returns true if everything is ok.
+   */
+  private async restore(): Promise<boolean> {
+    try {
+      const restoredAddresses = await this.restoreAddresses();
+      restoredAddresses.forEach(addr => this.persistAddressToCache(addr));
+      return true;
+    } catch (e) {
```

is the error should throw an error (and not return false)? @tiero

---

Yes. I think the try and catch should only handle errors (ie. API) so the user knows to retry, otherwise the false could be misleading.

Since this functions has side effect, we can infer the "true" by thefact we do not have errors